### ISSUE TITLE
Error Trailing characters without any trailing characters

### ISSUE THREAD

**Issue by hirohido** [serde-rs/json#944]
I am writing a UDP server and i am trying to de-serialize the data received. i get the error `trailing characters at line 1 column 8`
i print the string before de-serializing and it looks all right to me. I also tested it with an online json tester and it says valid json. but serde gives this error.

the string looks like this when printed `{"c":1}`

**Comment by dtolnay** [comment]
Please provide a compilable, runnable, minimal Rust program that reproduces this behavior.

**Comment by hirohido** [comment]
I found the problem. I used a very large buffer to receive the data from the socket. this buffer contained hundreds of 0s after the actual valid utf8. i printed the string as an array of bytes and i saw that there were many many 0s:

The equivalent of `{"c":1}` is:
[123, 34, 99, 34, 58, 32, 49, 32, 125]

But what i was unknowingly giving serde was:
[123, 34, 99, 34, 58, 32, 49, 32, 125, 0, 0, 0, 0, 0, 0, 0, 0, 0,]
with hundreds of more zeros.

so how i fixed this was:
`let data = str::from_utf8(&buffer).unwrap();`
`let trimmed_data = data.trim_end_matches(char::from(0));`

and now it works.
